internal/uvm/create_wcow.go

```diff
@@ -204,10 +155,138 @@ func CreateWCOW(ctx context.Context, opts *OptionsWCOW) (_ *UtilityVM, err error
+			return nil, err
+		}
+		opts.ID = g.String()
+	}
+
+	span.AddAttributes(trace.StringAttribute(logfields.UVMID, opts.ID))
+	log.G(ctx).WithField("options", fmt.Sprintf("%+v", opts)).Debug("uvm::CreateWCOW options")
+
+	uvm := &UtilityVM{
+		id:                      opts.ID,
+		owner:                   opts.Owner,
+		operatingSystem:         "windows",
+		scsiControllerCount:     1,
+		vsmbDirShares:           make(map[string]*VSMBShare),
+		vsmbFileShares:          make(map[string]*VSMBShare),
+		vpciDevices:             make(map[string]*VPCIDevice),
+		physicallyBacked:        !opts.AllowOvercommit,
+		devicesPhysicallyBacked: opts.FullyPhysicallyBacked,
+	}
+
+	defer func() {
+		if err != nil {
+			uvm.Close()
+		}
+	}()
+
+	if err := verifyOptions(ctx, opts); err != nil {
+		return nil, errors.Wrap(err, errBadUVMOpts.Error())
+	}
+
+	uvmFolder, err := uvmfolder.LocateUVMFolder(ctx, opts.LayerFolders)
+	if err != nil {
+		return nil, fmt.Errorf("failed to locate utility VM folder from layer folders: %s", err)
+	}
+
+	// TODO: BUGBUG Remove this. @jhowardmsft
+	//       It should be the responsiblity of the caller to do the creation and population.
+	//       - Update runhcs too (vm.go).
+	//       - Remove comment in function header
+	//       - Update tests that rely on this current behaviour.
+	// Create the RW scratch in the top-most layer folder, creating the folder if it doesn't already exist.
+	scratchFolder := opts.LayerFolders[len(opts.LayerFolders)-1]
+
+	// Create the directory if it doesn't exist
+	if _, err := os.Stat(scratchFolder); os.IsNotExist(err) {
+		if err := os.MkdirAll(scratchFolder, 0777); err != nil {
+			return nil, fmt.Errorf("failed to create utility VM scratch folder: %s", err)
+		}
+	}
+
+	doc, err := prepareConfigDoc(ctx, uvm, opts, uvmFolder)
+	if err != nil {
+		return nil, fmt.Errorf("error in preparing config doc: %s", err)
+	}
+
+	if !opts.IsClone {
+		// Create sandbox.vhdx in the scratch folder based on the template, granting the correct permissions to it
+		scratchPath := filepath.Join(scratchFolder, "sandbox.vhdx")
+		if _, err := os.Stat(scratchPath); os.IsNotExist(err) {
+			if err := wcow.CreateUVMScratch(ctx, uvmFolder, scratchFolder, uvm.id); err != nil {
+				return nil, fmt.Errorf("failed to create scratch: %s", err)
+			}
+		}
+
+		doc.VirtualMachine.Devices.Scsi = map[string]hcsschema.Scsi{
+			"0": {
+				Attachments: map[string]hcsschema.Attachment{
+					"0": {
+						Path:  scratchPath,
+						Type_: "VirtualDisk",
+					},
+				},
+			},
+		}
+
+		uvm.scsiLocations[0][0] = &SCSIMount{
+			vm:       uvm,
+			HostPath: doc.VirtualMachine.Devices.Scsi["0"].Attachments["0"].Path,
+			refCount: 1,
+		}
+	} else {
+		doc.VirtualMachine.RestoreState = &hcsschema.RestoreState{}
+		doc.VirtualMachine.RestoreState.TemplateSystemId = opts.TemplateConfig.UVMID
+
+		for _, cloneableResource := range opts.TemplateConfig.Resources {
+			_, err = cloneableResource.Clone(ctx, uvm, &cloneData{
```

Comment:
Do we need to do something with the returned `Resource` here? It seems weird to just be dropping it.

Reply:
Usually when we add a new resource to the uvm we store the resource struct (`VSMBShare`, `SCSIMount` etc) returned by the `Add*` call inside the uvm struct. But in this case that is done by the Clone call itself. So I think it is okay to drop the returned resource. We can change the Clone function definition to not return the resource at all if that sounds better.

Reply:
In that case, should `Clone` return a resource at all?

Reply:
Updated the definition to not return the resource.